forem/forem
# Push Notification multi app support

## What type of PR is this? (check all applicable)


## Description

Implements the multi app support for Forem instances mentioned in [RFC 68](https://github.com/forem/rfcs/pull/68).

What's included in this PR:
- New CRUD Admin page `/admin/consumer_apps`
- Offloaded some responsibilities from the `Device` model into the `ConsumerApp` model
   - They will store the credentials for each app (ConsumerApp) that will receive PNs
- New Query objects were created to fetch these models (found in `app/queries/consumer_apps`)

## Related Tickets & Documents

[RFC 68](https://github.com/forem/rfcs/pull/68)

## QA Instructions, Screenshots, Recordings

1. `bin/setup && bin/startup`
2. Log in with Admin user and visit `/admin/consumer_apps`
   - You should be able to see `com.forem.app` available already and you should not be able to modify it in any way
   - `com.forem.app` should appear as inactive (❌) because your local environment doesn't have the certificate in the designated ENV variable
   - If you add `export RPUSH_IOS_PEM="gibberish"` to your `.env` file and reboot it should now appear as active (✅)
3. Play around creating, editing and deleting Push Notification Targets
   - Make sure that if the Authentication Key is not available (empty) the target will appear as inactive (❌)

------------

![Screen Shot 2021-04-21 at 11 29 58](https://user-images.githubusercontent.com/6045239/115596071-ea9dc600-a294-11eb-90bf-318b653d8b37.png)

### UI accessibility concerns?

I'm not sure how we're checking for accessibility concerns in the Admin panel, specially considering the upcoming refactors to it. I'd be happy to address any request that doesn't block this PR for too long to make the UI more accessible.

## Added tests?


## [Forem core team only] How will this change be communicated?

- [x] This change does not need to be communicated, and this is why not: This is part of RFC 68 and it's still not completed yet. The moment it is shipped in its entirety it will be communicated appropriately.

## [optional] Are there any post deployment tasks we need to perform?

- [ ] Follow up PR to remove `app_bundle` from `devices` table and the `ignored_columns` from `Device` model
- [ ] Make sure `RPUSH_IOS_PEM` is available to all Forem Cloud instances

## [optional] What gif best describes this PR or how it makes you feel?

![Dr Frankenstein](https://media.giphy.com/media/l3vRlInF7QViJNOow/giphy.gif)
